File: app/src/main/java/com/example/taxconnect/RegisterActivity.java
package com.example.taxconnect;

import android.content.Intent;
import android.os.Bundle;
import android.net.Uri;
import android.text.TextUtils;
import android.view.View;
import android.widget.Toast;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.appcompat.app.AppCompatActivity;

import com.example.taxconnect.databinding.ActivityRegisterBinding;
import com.example.taxconnect.model.UserModel;
import com.example.taxconnect.services.CloudinaryHelper;

public class RegisterActivity extends AppCompatActivity {

    private ActivityRegisterBinding binding;
    private DataRepository repository;
    private String role = "CUSTOMER"; // Default
    
    private Uri selectedProfileImageUri;
    private Uri selectedIntroVideoUri;
    private String uploadedProfileImageUrl;
    private String uploadedIntroVideoUrl;

    private final ActivityResultLauncher<String> pickImageLauncher = registerForActivityResult(
            new ActivityResultContracts.GetContent(),
            uri -> {
                if (uri != null) {
                    selectedProfileImageUri = uri;
                    binding.ivProfilePreview.setImageURI(uri);
                    binding.btnUploadProfilePic.setText("Photo Selected");
                }
            }
    );

    private final ActivityResultLauncher<String> pickVideoLauncher = registerForActivityResult(
            new ActivityResultContracts.GetContent(),
            uri -> {
                if (uri != null) {
                    selectedIntroVideoUri = uri;
                    binding.ivVideoPreview.setVisibility(View.VISIBLE);
                    binding.tvVideoPlaceholder.setVisibility(View.GONE);
                    // Use Glide or just set a placeholder for video, or try to load thumbnail
                    // For now, just show the icon or text
                    binding.tvVideoPlaceholder.setText("Video Selected");
                    binding.tvVideoPlaceholder.setVisibility(View.VISIBLE);
                    binding.btnUploadIntroVideo.setText("Video Selected");
                }
            }
    );

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityRegisterBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        
        repository = DataRepository.getInstance();
        
        // Handle Intent Extra
        if (getIntent().hasExtra("ROLE")) {
            String intentRole = getIntent().getStringExtra("ROLE");
            if ("CA".equals(intentRole)) {
                binding.rbCA.setChecked(true);
                role = "CA";
                binding.layoutCAFields.setVisibility(View.VISIBLE);
                binding.layoutCustomerFields.setVisibility(View.GONE);
            } else {
                binding.rbCustomer.setChecked(true);
                role = "CUSTOMER";
                binding.layoutCAFields.setVisibility(View.GONE);
                binding.layoutCustomerFields.setVisibility(View.VISIBLE);
            }
        } else {
             // Default state is Customer checked
             binding.layoutCAFields.setVisibility(View.GONE);
             binding.layoutCustomerFields.setVisibility(View.VISIBLE);
        }
        
        setupListeners();
    }
    
    private void setupListeners() {
        // Setup Card Selection Listeners
        binding.cardClient.setOnClickListener(v -> selectRole("CUSTOMER"));
        binding.cardCA.setOnClickListener(v -> selectRole("CA"));

        binding.btnRegister.setOnClickListener(v -> registerUser());
        
        binding.tvLoginLink.setOnClickListener(v -> {
            startActivity(new Intent(RegisterActivity.this, LoginActivity.class));
            finish();
        });
        
        binding.btnUploadProfilePic.setOnClickListener(v -> pickImageLauncher.launch("image/*"));
        binding.btnUploadIntroVideo.setOnClickListener(v -> pickVideoLauncher.launch("video/*"));
    }

    private void selectRole(String selectedRole) {
        this.role = selectedRole;
        if ("CA".equals(selectedRole)) {
            binding.rbCA.setChecked(true);
            binding.rbCustomer.setChecked(false);
            binding.cardCA.setStrokeColor(getColor(R.color.primary));
            binding.cardClient.setStrokeColor(getColor(R.color.text_muted));
            
            // Update Text Colors
            ((android.widget.TextView)((android.widget.LinearLayout)binding.cardCA.getChildAt(0)).getChildAt(1)).setTextColor(getColor(R.color.primary));
            ((android.widget.ImageView)((android.widget.LinearLayout)binding.cardCA.getChildAt(0)).getChildAt(0)).setColorFilter(getColor(R.color.primary));
            
            ((android.widget.TextView)((android.widget.LinearLayout)binding.cardClient.getChildAt(0)).getChildAt(1)).setTextColor(getColor(R.color.text_muted));
            ((android.widget.ImageView)((android.widget.LinearLayout)binding.cardClient.getChildAt(0)).getChildAt(0)).setColorFilter(getColor(R.color.text_muted));

            binding.layoutCAFields.setVisibility(View.VISIBLE);
            binding.layoutCustomerFields.setVisibility(View.GONE);
        } else {
            binding.rbCustomer.setChecked(true);
            binding.rbCA.setChecked(false);
            binding.cardClient.setStrokeColor(getColor(R.color.primary));
            binding.cardCA.setStrokeColor(getColor(R.color.text_muted));
            
            // Update Text Colors
            ((android.widget.TextView)((android.widget.LinearLayout)binding.cardClient.getChildAt(0)).getChildAt(1)).setTextColor(getColor(R.color.primary));
            ((android.widget.ImageView)((android.widget.LinearLayout)binding.cardClient.getChildAt(0)).getChildAt(0)).setColorFilter(getColor(R.color.primary));
            
            ((android.widget.TextView)((android.widget.LinearLayout)binding.cardCA.getChildAt(0)).getChildAt(1)).setTextColor(getColor(R.color.text_muted));
            ((android.widget.ImageView)((android.widget.LinearLayout)binding.cardCA.getChildAt(0)).getChildAt(0)).setColorFilter(getColor(R.color.text_muted));

            binding.layoutCAFields.setVisibility(View.GONE);
            binding.layoutCustomerFields.setVisibility(View.VISIBLE);
        }
    }
    
    private void registerUser() {
        String name = binding.etName.getText().toString().trim();
        String email = binding.etEmail.getText().toString().trim();
        String city = binding.etCity.getText().toString().trim();
        String password = binding.etPassword.getText().toString().trim();
        
        if (TextUtils.isEmpty(name) || TextUtils.isEmpty(email) || TextUtils.isEmpty(city) || TextUtils.isEmpty(password)) {
            Toast.makeText(this, "Please fill all basic fields including City", Toast.LENGTH_SHORT).show();
            return;
        }
        
        UserModel user = new UserModel(null, email, name, role);
        user.setCity(city);
        
        if ("CA".equals(role)) {
            String caNumber = binding.etCaNumber.getText().toString().trim();
            String experience = binding.etExperience.getText().toString().trim();
            String specialization = binding.etSpecialization.getText().toString().trim();
            String minCharges = binding.etMinCharges.getText().toString().trim();
            
            if (TextUtils.isEmpty(caNumber) || TextUtils.isEmpty(experience) || 
                TextUtils.isEmpty(specialization) || TextUtils.isEmpty(minCharges)) {
                Toast.makeText(this, "Please fill all CA fields", Toast.LENGTH_SHORT).show();
                return;
            }

            if (selectedProfileImageUri == null || selectedIntroVideoUri == null) {
                Toast.makeText(this, "Please select both profile picture and intro video", Toast.LENGTH_SHORT).show();
                return;
            }
            
            user.setCaNumber(caNumber);
            user.setExperience(experience);
            user.setSpecialization(specialization);
            user.setMinCharges(minCharges);
            
            uploadMediaAndRegister(email, password, user);
        } else {
            performRegistration(email, password, user);
        }
    }

    private void uploadMediaAndRegister(String email, String password, UserModel user) {
        binding.btnRegister.setEnabled(false);
        binding.btnRegister.setText("Uploading Media...");
        binding.progressBarMedia.setVisibility(View.VISIBLE);

        CloudinaryHelper.uploadMedia(this, selectedProfileImageUri, new CloudinaryHelper.ImageUploadCallback() {
            @Override
            public void onSuccess(String url) {
                uploadedProfileImageUrl = url;
                user.setProfileImageUrl(uploadedProfileImageUrl);
                
                CloudinaryHelper.uploadMedia(RegisterActivity.this, selectedIntroVideoUri, new CloudinaryHelper.ImageUploadCallback() {
                    @Override
                    public void onSuccess(String videoUrl) {
                        uploadedIntroVideoUrl = videoUrl;
                        user.setIntroVideoUrl(uploadedIntroVideoUrl);
                        
                        runOnUiThread(() -> {
                            binding.progressBarMedia.setVisibility(View.GONE);
                            performRegistration(email, password, user);
                        });
                    }

                    @Override
                    public void onError(String error) {
                        runOnUiThread(() -> {
                            binding.btnRegister.setEnabled(true);
                            binding.btnRegister.setText("Create Account");
                            binding.progressBarMedia.setVisibility(View.GONE);
                            Toast.makeText(RegisterActivity.this, "Video Upload Failed: " + error, Toast.LENGTH_SHORT).show();
                        });
                    }
                });
            }

            @Override
            public void onError(String error) {
                runOnUiThread(() -> {
                    binding.btnRegister.setEnabled(true);
                    binding.btnRegister.setText("Create Account");
                    binding.progressBarMedia.setVisibility(View.GONE);
                    Toast.makeText(RegisterActivity.this, "Image Upload Failed: " + error, Toast.LENGTH_SHORT).show();
                });
            }
        });
    }

    private void performRegistration(String email, String password, UserModel user) {
        binding.btnRegister.setEnabled(false);
        binding.btnRegister.setText("Registering...");
        
        repository.registerUser(email, password, user, new DataRepository.DataCallback<Void>() {
            @Override
            public void onSuccess(Void data) {
                Toast.makeText(RegisterActivity.this, "Registration Successful!", Toast.LENGTH_SHORT).show();
                navigateToDashboard();
            }

            @Override
            public void onError(String error) {
                binding.btnRegister.setEnabled(true);
                binding.btnRegister.setText("Create Account");
                Toast.makeText(RegisterActivity.this, "Error: " + error, Toast.LENGTH_LONG).show();
            }
        });
    }
    
    private void navigateToDashboard() {
        Intent intent;
        if ("CA".equals(role)) {
            intent = new Intent(this, CADashboardActivity.class);
        } else {
            intent = new Intent(this, HomeActivity.class);
        }
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
        startActivity(intent);
    }
}

File: app/src/main/java/com/example/taxconnect/features/auth/RegisterActivity.kt
package com.example.taxconnect.features.auth

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.text.TextUtils
import android.util.Patterns
import android.view.View
import android.widget.LinearLayout
import android.widget.TextView
import android.widget.ImageView
import android.view.LayoutInflater
import androidx.activity.result.contract.ActivityResultContracts
import com.example.taxconnect.R
import com.example.taxconnect.core.base.BaseActivity
import com.example.taxconnect.data.repositories.DataRepository
import com.example.taxconnect.data.repositories.DataRepository.DataCallback
import com.example.taxconnect.databinding.ActivityRegisterBinding
import com.example.taxconnect.data.models.UserModel
import com.example.taxconnect.data.services.CloudinaryHelper
import com.example.taxconnect.core.network.NetworkUtils
import com.example.taxconnect.features.dashboard.CADashboardActivity
import com.example.taxconnect.features.home.HomeActivity

class RegisterActivity : BaseActivity<ActivityRegisterBinding>() {

    override val bindingInflater: (LayoutInflater) -> ActivityRegisterBinding = ActivityRegisterBinding::inflate
    private lateinit var repository: DataRepository
    private var role = "CUSTOMER" // Default

    private var selectedProfileImageUri: Uri? = null
    private var selectedIntroVideoUri: Uri? = null
    private var uploadedProfileImageUrl: String? = null
    private var uploadedIntroVideoUrl: String? = null

    private val pickImageLauncher = registerForActivityResult(ActivityResultContracts.GetContent()) { uri ->
        if (uri != null) {
            selectedProfileImageUri = uri
            binding.ivProfilePreview.setImageURI(uri)
            binding.btnUploadProfilePic.text = getString(R.string.photo_selected)
        }
    }

    private val pickVideoLauncher = registerForActivityResult(ActivityResultContracts.GetContent()) { uri ->
        if (uri != null) {
            selectedIntroVideoUri = uri
            binding.ivVideoPreview.visibility = View.VISIBLE
            binding.tvVideoPlaceholder.visibility = View.GONE
            binding.tvVideoPlaceholder.text = getString(R.string.video_selected)
            binding.tvVideoPlaceholder.visibility = View.VISIBLE
            binding.btnUploadIntroVideo.text = getString(R.string.video_selected)
        }
    }

    override fun initViews() {
        repository = DataRepository.getInstance()

        if (intent.hasExtra("ROLE")) {
            val intentRole = intent.getStringExtra("ROLE")
            if ("CA" == intentRole) {
                binding.rbCA.isChecked = true
                role = "CA"
                binding.layoutCAFields.visibility = View.VISIBLE
                binding.layoutCustomerFields.visibility = View.GONE
            } else {
                binding.rbCustomer.isChecked = true
                role = "CUSTOMER"
                binding.layoutCAFields.visibility = View.GONE
                binding.layoutCustomerFields.visibility = View.VISIBLE
            }
        } else {
            binding.layoutCAFields.visibility = View.GONE
            binding.layoutCustomerFields.visibility = View.VISIBLE
        }
    }

    override fun observeViewModel() {
        // No ViewModel to observe
    }

    override fun setupListeners() {
        binding.cardClient.setOnClickListener { selectRole("CUSTOMER") }
        binding.cardCA.setOnClickListener { selectRole("CA") }

        binding.btnRegister.setOnClickListener { registerUser() }

        binding.tvLoginLink.setOnClickListener {
            startActivity(Intent(this@RegisterActivity, LoginActivity::class.java))
            finish()
        }

        binding.btnUploadProfilePic.setOnClickListener { pickImageLauncher.launch("image/*") }
        binding.btnUploadIntroVideo.setOnClickListener { pickVideoLauncher.launch("video/*") }
    }

    private fun selectRole(selectedRole: String) {
        this.role = selectedRole
        if ("CA" == selectedRole) {
            binding.rbCA.isChecked = true
            binding.rbCustomer.isChecked = false
            binding.cardCA.strokeColor = getColor(R.color.primary)
            binding.cardClient.strokeColor = getColor(R.color.text_muted)

            ((binding.cardCA.getChildAt(0) as LinearLayout).getChildAt(1) as TextView).setTextColor(getColor(R.color.primary))
            ((binding.cardCA.getChildAt(0) as LinearLayout).getChildAt(0) as ImageView).setColorFilter(getColor(R.color.primary))

            ((binding.cardClient.getChildAt(0) as LinearLayout).getChildAt(1) as TextView).setTextColor(getColor(R.color.text_muted))
            ((binding.cardClient.getChildAt(0) as LinearLayout).getChildAt(0) as ImageView).setColorFilter(getColor(R.color.text_muted))

            binding.layoutCAFields.visibility = View.VISIBLE
            binding.layoutCustomerFields.visibility = View.GONE
        } else {
            binding.rbCustomer.isChecked = true
            binding.rbCA.isChecked = false
            binding.cardClient.strokeColor = getColor(R.color.primary)
            binding.cardCA.strokeColor = getColor(R.color.text_muted)

            ((binding.cardClient.getChildAt(0) as LinearLayout).getChildAt(1) as TextView).setTextColor(getColor(R.color.primary))
            ((binding.cardClient.getChildAt(0) as LinearLayout).getChildAt(0) as ImageView).setColorFilter(getColor(R.color.primary))

            ((binding.cardCA.getChildAt(0) as LinearLayout).getChildAt(1) as TextView).setTextColor(getColor(R.color.text_muted))
            ((binding.cardCA.getChildAt(0) as LinearLayout).getChildAt(0) as ImageView).setColorFilter(getColor(R.color.text_muted))

            binding.layoutCAFields.visibility = View.GONE
            binding.layoutCustomerFields.visibility = View.VISIBLE
        }
    }

    private fun registerUser() {
        if (!NetworkUtils.isNetworkAvailable(this)) {
            showToast(getString(R.string.no_internet))
            return
        }

        val name = binding.etName.text.toString().trim()
        val email = binding.etEmail.text.toString().trim()
        val city = binding.etCity.text.toString().trim()
        val password = binding.etPassword.text.toString().trim()

        if (TextUtils.isEmpty(name)) {
            binding.etName.error = getString(R.string.name_required)
            binding.etName.requestFocus()
            return
        }

        if (TextUtils.isEmpty(email)) {
            binding.etEmail.error = getString(R.string.email_required)
            binding.etEmail.requestFocus()
            return
        }

        if (!Patterns.EMAIL_ADDRESS.matcher(email).matches()) {
            binding.etEmail.error = getString(R.string.invalid_email)
            binding.etEmail.requestFocus()
            return
        }

        if (TextUtils.isEmpty(city)) {
            binding.etCity.error = getString(R.string.city_required)
            binding.etCity.requestFocus()
            return
        }

        if (TextUtils.isEmpty(password)) {
            binding.etPassword.error = getString(R.string.password_required)
            binding.etPassword.requestFocus()
            return
        }

        if (password.length < 6) {
            binding.etPassword.error = getString(R.string.password_min_length)
            binding.etPassword.requestFocus()
            return
        }

        val user = UserModel(null, email, name, role)
        user.city = city

        if ("CA" == role) {
            val caNumber = binding.etCaNumber.text.toString().trim()
            val experience = binding.etExperience.text.toString().trim()
            val specialization = binding.etSpecialization.text.toString().trim()
            val minCharges = binding.etMinCharges.text.toString().trim()

            if (TextUtils.isEmpty(caNumber)) {
                binding.etCaNumber.error = getString(R.string.ca_number_required)
                binding.etCaNumber.requestFocus()
                return
            }
            if (caNumber.length < 6) {
                binding.etCaNumber.error = getString(R.string.ca_number_min_length)
                binding.etCaNumber.requestFocus()
                return
            }
            if (TextUtils.isEmpty(experience)) {
                binding.etExperience.error = getString(R.string.experience_required)
                binding.etExperience.requestFocus()
                return
            }
            if (TextUtils.isEmpty(specialization)) {
                binding.etSpecialization.error = getString(R.string.specialization_required)
                binding.etSpecialization.requestFocus()
                return
            }
            if (TextUtils.isEmpty(minCharges)) {
                binding.etMinCharges.error = getString(R.string.min_charges_required)
                binding.etMinCharges.requestFocus()
                return
            }

            if (selectedProfileImageUri == null) {
                showToast(getString(R.string.select_profile_pic))
                binding.btnUploadProfilePic.requestFocus()
                return
            }
            if (selectedIntroVideoUri == null) {
                showToast(getString(R.string.select_intro_video))
                binding.btnUploadIntroVideo.requestFocus()
                return
            }

            user.caNumber = caNumber
            user.experience = experience
            user.specialization = specialization
            user.minCharges = minCharges

            uploadMediaAndRegister(email, password, user)
        } else {
            performRegistration(email, password, user)
        }
    }

    private fun uploadMediaAndRegister(email: String, password: String, user: UserModel) {
        binding.btnRegister.isEnabled = false
        binding.btnRegister.text = getString(R.string.uploading_media)
        binding.progressBarMedia.visibility = View.VISIBLE

        CloudinaryHelper.uploadMedia(this, selectedProfileImageUri!!, object : CloudinaryHelper.ImageUploadCallback {
            override fun onSuccess(data: String?) {
                uploadedProfileImageUrl = data ?: ""
                user.profileImageUrl = uploadedProfileImageUrl

                CloudinaryHelper.uploadMedia(this@RegisterActivity, selectedIntroVideoUri!!, object : CloudinaryHelper.ImageUploadCallback {
                    override fun onSuccess(data: String?) {
                        uploadedIntroVideoUrl = data ?: ""
                        user.introVideoUrl = uploadedIntroVideoUrl

                        runOnUiThread {
                            binding.progressBarMedia.visibility = View.GONE
                            performRegistration(email, password, user)
                        }
                    }

                    override fun onError(error: String?) {
                        runOnUiThread {
                            binding.btnRegister.isEnabled = true
                            binding.btnRegister.text = getString(R.string.create_account)
                            binding.progressBarMedia.visibility = View.GONE
                            showToast(getString(R.string.video_upload_failed, error))
                        }
                    }
                })
            }

            override fun onError(error: String?) {
                runOnUiThread {
                    binding.btnRegister.isEnabled = true
                    binding.btnRegister.text = getString(R.string.create_account)
                    binding.progressBarMedia.visibility = View.GONE
                    showToast(getString(R.string.image_upload_failed, error))
                }
            }
        })
    }

    private fun performRegistration(email: String, password: String, user: UserModel) {
        binding.btnRegister.isEnabled = false
        binding.btnRegister.text = getString(R.string.registering)

        repository.registerUser(email, password, user, object : DataCallback<Void?> {
            override fun onSuccess(data: Void?) {
                showToast(getString(R.string.registration_successful))
                navigateToDashboard()
            }

            override fun onError(error: String?) {
                binding.btnRegister.isEnabled = true
                binding.btnRegister.text = getString(R.string.create_account)
                showToast(getString(R.string.error_with_message, error))
            }
        })
    }

    private fun navigateToDashboard() {
        val intent = if ("CA" == role) {
            Intent(this, CADashboardActivity::class.java)
        } else {
            Intent(this, HomeActivity::class.java)
        }
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK)
        startActivity(intent)
    }
}
